오픈마켓 [STEP 1] 요한, 줄라이

July911 opened the pull request:
안녕하세요 콘! (@protocorn93)
요한(@yohanblessyou) 과 줄라이입니다. 
새해 복 많이 받으세요 👍 👍 👍  😄 😄 
 너무 늦어져서 죄송합니다.  
데이터 통신이 처음이라 너무 많이 헤맨것도 한 몫 하는 것 같습니다... 많은 배움을 얻고싶습니다..
이번 프로젝트는 유난히 헷갈리는게 많아서 궁금한점만...적었습니다 ! 
감사합니다 😄 

## 궁금한 점

### 🐥  1. MockURLSession을 만들어 테스트하려는 대상 기능이 무엇인지 궁금합니다 🐥 

현재 구현된 코드는 "URLSession으로 JSON 파일을 불러와 ParsingType 인스턴스로 변환"하는 코드입니다. 이후 전체 코드에서 어떤 부분을 테스트해야할지 고민하던 중 의문이 들었습니다

전체 코드를 기능 단위로 아래와 같이 나누어 보았습니다

1. URLRequest 생성 
2. 서버로 요청을 전송 (URLSession)
3. 서버로부터 응답을 받음 (URLSession)
4. 응답 data가 JSON이라면 ParsingType 인스턴스로 디코딩

의문이 드는 점은 많은 사람들이 MockURLSession을 만들어 테스트하고 있는데, URLSession을 Mock으로 대체하여 테스트하려는 대상 기능이 무엇인지 의문이 들었습니다 

(MockURLSession을 만드는 것이 프로젝트의 요구사항은 아닙니다)

현재 URLSession을 Mock으로 대체하는 테스트 중이므로 2번,3번은 자연히 대상이 아니고, 1번도 우리가 생성한 Request가 유효한지는 결국 서버의 내부 알고리즘에 따른 것이므로 Mock에 임의로 구현할 수 있는 부분이 아닌 것 같습니다. 마지막으로 4번 또한 디코딩 테스트만 할 것이라면 MockURLSession이 필요하진 않은 것 같습니다

결국 서버를 통해야만 1~3번 기능 테스트가 가능하고 4번은 MockURLSession이 필요없는 것 같아 URLSession을 Mock으로 만드는건 무의미한 것 같은데.. 워낙 많은 사람들이 만들어 테스트하고 있어 혹시 저희가 오해하고 있는 부분이 있는지 궁금합니다!


### 🐥  2. static 프로퍼티에 관하여 🐥 

![](https://i.imgur.com/A5jo82l.jpg)

각각의 Requester 들은 각자 baseURLString 을 가지고 있습니다. 한번 인스턴스를 생성할 때 마다 동일한 baseURL을 가지고 있어서, 메모리 낭비라는 생각을 하였습니다. 또한 오픈마켓이라는 어플이라고 가정했을 때, 데이터를 주고 받는일이 빈번하게 일어날 것 같아 static 을 사용하여 전역프로퍼티로 만들어주었습니다. 하지만 전역으로 만들어주는것을 지양하는 것이 좋다고 배웠는데, 콘이라면 어떻게 하실지 궁금합니다.  


### 🐥  3. DataTask 비동기에 관하여 🐥 
![](https://i.imgur.com/K7hEie0.jpg)
![](https://i.imgur.com/Zax4Zsm.jpg)

@escaping 을 사용한 탈출 클로저를 만들어주었을 때, 활동학습 때 들었던 바로는 비동기로 처리된다고 배웠는데, 저희가 만든 mock dataTask 는 비동기로 작동하지 않는다는것을 배웠습니다. 공식문서로 확인한 결과 엔딩클로저가 비동기로 작업된다는 내용이 없어서, 어떤게 맞는건지 헷갈리기 시작했습니다.
애플이 만든 URLSession의 dataTask는 자동으로 비동기 처리가 되는걸로 확인했는데, 엔딩클로저 때문이 아니라 그냥 main.async 로 만들어진건지 궁금합니다. 
### 🐥  4. 파일분리에 관하여 🐥 

<img src="https://i.imgur.com/WdKsIDz.jpg" width="30%">

각각의 인코딩과 디코딩을 위한 객체를 따로 분리해서 만들어주었습니다. 그리고 url을 가지고있고, request를 생성해주는 requester를 각각 만들어주었습니다. 뷰나 뷰 컨트롤러에서 해주는 작업이 아니여서 모델안으로 파일을 넣어주었는데, 모델이라고 보기에도 (데이터를 주고받는 부분임에도) 어색하다고 느껴졌습니다. 네트워킹이라는 파일을 따로 분류해서 넣어주는것이 맞는건지, 아니면 데이터통신을 처음해봐서 어색하게 느껴져서 그런건지 궁금합니다. -> 제 생각은 네트워킹이라는 타입을 따로 만들고, requester가 채택한 requestable을 프로퍼티로 가지고 있는 APICaller라는 객체만 모델에 있는것이 맞다고 생각합니다.

youngwoo-jang committed:
feat: URLSession Extension 구현 #1

July911 committed:
feat: NetWorkingType 초기구현 #1

youngwoo-jang committed:
feat: HTTPRequest 타입 구현 #1

July911 committed:
test: Mock URLSession 타입 구현 #2

July911 committed:
feat: URLSessionProtocol 생성 #1

- Extension 으로 request() 메서드 구현
- Test를 위한 StubURLSession에 URLSessionProtocol 채택

youngwoo-jang committed:
test: HTTP GET 메서드 테스트 코드 구현 #2

- URLSession이 URLSessionProtocol 채택

July911 committed:
refactor: API별 타입 분리 #1

- ProductEntry,ProductModify,ProductSecretAsk

youngwoo-jang committed:
refactor: API별 타입 분리 #1

- ProductDelete, ProductDetailAsk, ProductListAsk

July911 committed:
feat: 각 기능별 Requester 생성 #3

July911 committed:
feat: Requester 별 HTTP Method 및 URL 구현 #3

youngwoo-jang committed:
feat: Model 프로토콜 생성 #3

- JSONParsable 프로토콜 생성
- URLSessionProtocol의 request 메서드 구조 변경

July911 committed:
test: MockURLSession테스트 코드 작성 #2

youngwoo-jang committed:
fix: MockURLSession 이니셜라이져 코드 수정 #2

youngwoo-jang committed:
refactor: 파일 정리 #1

protocorn93 commented on OpenMarket/OpenMarket/Model/HttpMethod.swift:
```diff
@@ -0,0 +1,6 @@
+enum HttpMethod: String {
```
보통 HTTP와 같은 약어가 앞에 오게 된다면 모두 대문자로 표기해주는게 보편적인 스타일인 것 같아요.

July911 replied:
```swift 
enum HTTPMethod: String {
``` 
바꿔주었습니다 !

protocorn93 commented on OpenMarket/OpenMarket/Model/ParsingType/ProductDetailAsk.swift:
```diff
@@ -0,0 +1,55 @@
+import Foundation
+
+enum ProductDetailAsk {
+    struct Response: Decodable {
```
저는 개인적으로 이렇게 inner type으로 물리적인 계층을 나누어 사용하는게 좋더라구요. 다만 이게 계속 중첩된다면 최하단 타입을 직접 사용하는 곳에선 꽤나 길어진 타입명을 사용할텐데요. 그럴 땐 `typealias`를 사용하는 것도 좋은 해결방법이 될 수 있을 것 같아요

July911 replied:
좋은 방법을 알려주셔서 감사합니다 ! 단지 지금 계층 구조에서는 오히려 타입명을 명시해주는게 조금 더 직관적이라 판단이 들어 콘이 알려주신 방법만 숙지하고 현재 상태를 유지하기로 했습니다 ! 
추후 코드가 길어지고 계층구조가 중첩된다면 사용해보겠습니다 감사합니다 :)

protocorn93 commented on OpenMarket/OpenMarket/Model/ParsingType/ProductListAsk.swift:
```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+enum ProductListAsk {
+    struct Response: Decodable {
```
용도에 상관없이 `Encodable`을 채택하지 않고 딱 사용하는 `Decodable` 프로토콜만 채택한 점 훌륭합니다! 👏

protocorn93 commented on OpenMarket/OpenMarket/Model/Protocol/JSONParsable.swift:
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+protocol JSONParsable { }
```
`JSONParsable` 프로토콜에 메서드를 선언해야 하지 않을까요? 프로토콜 메서드가 아닌 기본 구현 메서드는 `JSONParsable`을 채택해도 이를 대체할 수가 없을 것 같아요. 

우리가 추상화를 하고, 이를 구현체에 맞게 코드를 작성해서 얻을 수 있는 장점은 "대체가능성"인데요. 현재 코드라면 구현체를 대체할 수 없을 것으로 보이네요.

protocorn93 replied:
파싱이라는 행위를 추상화하고,  `JSONDecoder`를 사용할지, 기타 서드파티 라이브러리를 통해서 파싱을 할지, 등은 구현체에서 담당하는게 좋을 것 같아요.

youngwoo-jang replied:
사실.. `JSONParsable`은 언급해주신 대체가능성을 구현하다가 실패하고.. "이 프로토콜을 채택한 타입은 JSON 포맷의 Response를 받는다"를 표현하는 정도의 기능을 위해 남아 있었습니다..

원래 목표했던 최종 형태는 아래와 같이, Requester(요청자 타입)가 자신이 받을 응답 JSON의 Decoding타입을 알고 있도록 하여 decode()의 파라미터로 넣을 필요가 없도록 만드는 것이었습니다

```swift
// Before
let requester = ProductListAskRequester()
let products = try? JSONDecoder().decode(requester.data, ProductListAsk.Response.self)

//After
let requester = ProductListAskRequester()
let products = try? requester.decode(requester.data)
```

---

그런데, 오늘 리팩터링 과정에서 `associatedType`을 통해 어느정도 대체가능성을 구현한 것 같습니다!
아래와 같이, 각 Requester가 DecodingType을 typealias로 정의하도록 하고, decodingtype을 파라미터로 받지 않는 `decode()`를 기본구현으로 제공하였습니다

```swift
protocol JSONResponseDecodable {
    associatedtype DecodingType: Decodable
}

extension JSONResponseDecodable {
    func decode(data: Data) -> DecodingType? {
        return try? JSONDecoder().decode(DecodingType.self, from: data)
    }
}
```

protocorn93 replied:
아하! 그런 의도셨군요! 그렇다면 `associatedType`을 적절히 잘 사용하신 것 같아요! 👍

protocorn93 commented on OpenMarket/OpenMarket/Model/Requester/HealthCheckerRequester.swift:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+struct HealthCheckerRequester: Requestable {
+    private static var baseURLString: String = "https://market-training.yagom-academy.kr/healthChecker"
```
`static`으로 굳이 가져갈 필요는 없어보입니다. 이를 사용하는 곳도 인스턴스 생성후 접근이 가능한 `get-only` 프로퍼티들이기에 `private let`이 좀 더 적절할 것 같아요

youngwoo-jang replied:
동의합니다! 기존엔 모든 인스턴스가 같은 값을 가지므로 하나의 메모리를 공유하게끔 `static`으로 선언하였었습니다

하지만, 어차피 인스턴스가 살아있는 동안에만 사용되므로 인스턴스 해제 시점에 함께 해제될 수 있도록 `private let`으로 변경하였습니다

protocorn93 commented on OpenMarket/OpenMarketTests/StubNetworkingType.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+class StubURLSessionDataTask: URLSessionDataTask {
```
기본 `URLSessionDataTask`를 상속받아 `resume`을 오버라이딩해서 호출 여부 검증을 하는데 사용한 점 매우 훌륭합니다!! 👏

protocorn93 commented on OpenMarket/OpenMarketTests/ParsingTests.swift:
```diff
@@ -0,0 +1,17 @@
+
+class ParsingTests: XCTestCase {
+
+    func test_ProductPageHTTPResponse가_정상적으로_디코딩되는지() throws {
```
우리가 "파싱"이라는 행위에서 검증해야하는 것은 input을 넣었을 때 기대하는 output이 나오느냐 일 것 같아요. (사실 모든 테스트 목적이 그렇긴하죠) 

그래서 우리가 여기서 테스트를 해볼 수 있는 것은 `JSONParsable`을 채택해서 `JSONDeocder`를 사용해 파싱을 구현한 구현체가 될 것 같아요. 

그리고 이를 사용하는 다른 객체를 테스트할 때는 `JSONParsable`을 채택한 Mock Parser를 주입해주어 실제 구현체를 대체할 수 있겠죠!

youngwoo-jang replied:
사실 이 부분을 작성하면서 의도했던 테스트 대상은 `ParsingType`들을 잘 정의했는가? 였어요. (ex. 프로퍼티 하나가 빠져 디코딩이 안된다던지)

그래서인지 작성 당시에는 `JSONParsable`에 대해서는 생각을 못했었던것 같아요. 말씀해주신 것처럼 `JSONParsable`이라는 프로토콜이 있기에 "파싱"테스트라고 하면 테스트 대상은 `JSONParsable`의 구현체가 되어야 할 것 같습니다! 

URLSession 외의 객체를 테스트하기 위해 MockURLSession을 만들었던 것처럼, Parser도 `JSONParsable`을 채택하는 Mock을 만들어 주입해볼 수 있겠네요!

protocorn93 commented on OpenMarket/OpenMarketTests/StubNetworkingType.swift:
```diff
@@ -0,0 +1,45 @@
+
+@testable import OpenMarket
+
+class StubURLSession: URLSessionProtocol {
```
목적에 따라 다양한 테스트 더블이 존재하는데요. Spy, Stub, Mock 등등.
이들에 대한 학습을 해보셨다면 간단하게 이해한 내용을 저에게 설명해주실 수 있으실까요?

youngwoo-jang replied:
우선, 각 테스트 더블 간 역할이 아주 명확하게 나뉘어 있지는 않다고 느꼈어요. 제가 이해한 바를 토대로 정리보겠습니다! 

- Stub
Mock에 비해 실제 객체의 더 작은 부분만을 구현합니다. 또한, 테스트 결과를 객체의 "상태"를 확인하여 결정하는 "상태검증 테스트"에 주로 사용됩니다. 예로, 현재 저희 코드의 `StubURLSession`처럼 호출되면 정해진 값(HTTP status code 등)을 반환하는 테스트 더블을 만들어야 하는 경우 Stub이 사용된다고 파악했습니다

- Mock
실제 객체와 가장 유사한 수준으로 구현됩니다. Stub과는 달리 객체의 어떤 메서드가 호출되었는지 등의 "행위검증 테스트"에 사용됩니다. 

---

- Dummy
가장 기본적인 테스트 더블로, 어떠한 기능도 구현되어 있지 않습니다. 테스트 코드에서 Dummy 객체에게는 어떠한 기능을 하길 바라지 않고 단순히 객체를 만들고, 넘기는 정도의 목적만 필요할 때 사용될 것 같습니다

- Fake
실제 객체의 로직을 단순화하여 구현한 테스트 더블을 말합니다. 실제 로직과는 차이가 있기에 실제 앱에서는 제대로 동작하지 않음

- Spy
기본적으로 Stub의 "상태검증 테스트"라는 목적을 가지면서, Mock의 "행위검증 테스트"를 위한 일부 로직(메서드 호출횟수를 기록한다던지)이 포함된 테스트 더블로 이해하였습니다.

protocorn93 replied:
간결하게 각각의 차이점을 너무 잘 설명해주셨어요! 종종 Stub, Mock을 혼용하여 사용하곤하는데 테스트 더블의 개념뿐만 아니라 다른 개념들도 정확하게 용도를 이해하고 사용하는 것이 코드리뷰나 의사소통에 있어서도 보다 명확한 의도를 전달할 수 있어요!

protocorn93 commented on OpenMarket/OpenMarket/Model/Protocol/Requestable.swift:
```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+protocol Requestable {
+    var request: URLRequest? { get }
```
`Requestable`은 api endpoint에 대한 정보를 추상화한 용도로 사용하고 이를 `URLSessionProtocol` 구현체 내부에서 사용해 `URLRequest`를 생성하는건 어떨까요? 

그렇게 되면 `URLRequest`를 활용한 네트워크 호출이 아니더라도 `Requestable`을 활용할 수 있을 것 같아요.

July911 replied:
```swift
protocol Requestable {
    var url: URL? { get }
    var httpMethod: HTTPMethod { get }
    var httpBody: Data? { get }
    var headerFields: [String: String]? { get }
}
```
URLRequest 만 할 수 있다는 생각이 잘못되었다는걸 알려주셔서 감사합니다. 
Requestable 프로토콜을 채택하면, 기본적으로 정보 등을 담을 수 있게 설계하였고, URLSessionProtocol 안으로 

```swift
func createURLRequest(requester: Requestable) throws -> URLRequest {
``` 
메서드를 빼주어, 리퀘스터만 받으면 그 안에서 정보를 받아서 URLRequest를 만들게 설계하였습니다.
또한 body와 header 가 없는 요청이 있을 수 있어서 옵셔널로 처리해주었습니다 !

protocorn93 commented on OpenMarket/OpenMarket/Model/Requester/ProductModifyRequester.swift:
```diff
@@ -0,0 +1,31 @@
+        var request = URLRequest(url: url)
+        request.httpMethod = Self.httpMethod.rawValue
+        
+        //TODO: need to add Body
```
`Requestable`에 남긴 피드백을 활용한다면 body parameter 역시 `[String: Any]`로 프로토콜에 포함시킬 수 있지 않을까 싶어요!

July911 replied:
```swift
  var headerFields: [String: String]? { get }
``` 
포함시켜주었습니다. 다만 any타입이 아니라, 현재는 String 만 받아오는 상황이라 Stringd으로 만들어 주었습니다 !

protocorn93 reviewed (commented):
두 분 모두 수고하셨습니다!! 쉽지 않은 주제인데 훌륭하게 구현해주셨어요! 

## 궁금한 점 

> 🐥 1. MockURLSession을 만들어 테스트하려는 대상 기능이 무엇인지 궁금합니다 

제 생각을 결론부터 말씀드리자면 네트워크를 호출할 때는 `URLSessionProtocol`을 채택한 객체를 생성해 직접 호출하지 않을거기 때문이에요. 이를 활용하는 Provider, Repository 등을 사용해서 네트워크 호출을 할거기 때문입니다. 

Provider, Repository 등의 객체들의 인터페이스는 더욱 단순할거에요
```swift
protocol RequestValueRepositoryProtocol { 
  func requestValues(someParameter: String, completion: ([Values], Error))
}
```
이런 Provider나 Repostiory를 사용하는 곳에선 이 객체가 로컬 데이터베이스에서 가져오는지, 파일에서 가져오는지, 네트워크에서 가져오는지 신경 쓸 필요가 없어요. 그저 "어딘가에 있는 데이터를 가져다주는 객체" 정도로 인식하고 사용합니다. Provider나 Repository에선 받아온 모델(`Decodable`을 채택한 모델)을 비즈니스 로직에서 사용되는 모델로의 전환 작업등을 담당하겠죠!

이렇게 계층을 하나 더 추가하게 되면 앱의 정책이 네트워크 호출에서 파일 저장 혹은 데이터베이스로의 저장으로 변경되거나 사용하는 서드파티 라이브러리가 변경되어도 (Core Data, Realm, Alamofire, AFNetworking 등등) 비즈니스 로직을 담당하는 코드나 뷰 컨트롤러의 코드 변경없이 갈아끼우기만하면 되기 때문에 변경에 보다 유연하게 대응할 수 있죠! 

`MockURLSession`을 만들어 테스트 하는 이유도 Provider, Repository를 테스트할 때 주입해주어 input과 output을 비교할 수 있기 때문입니다. Repository 패턴과 클린 아키텍처에 대해 간단하게 소개해주는 글들을 학습해보시면 좀 더 와닿으실 수 있을거에요! 

> 🐥 2. static 프로퍼티에 관하여 

네~ 이해했습니다. 하지만 저는 `ProductPostRequester` 객체를 생성하고 네트워크를 호출하는 과정에만 존재해야하는 데이터로 생각을 했어요. 그렇기 때문에 저는 이 부분에선 `static`을 쓰지 않는 것이 좋지 않나 라는 생각을합니다. 인스턴스의 라이프사이클 관점에서 바라보는 것 같아요. 

> 🐥 3. DataTask 비동기에 관하여 

네~ 그럼요! `URLSession`의 `dataTask`의 구현은 내부적으로 비동기로 처리되어 @escaping 클로저를 통해 결과를 전달받습니다. 메서드의 인터페이스와 비동기/동기는 무관합니다. 내부 구현을 어떻게 하느냐에 따라 다르겠죠! 현재 코드에서 Mock이 동기로 동작하는 정상 상황이라고 보여집니다. 

> 🐥 4. 파일분리에 관하여 

이렇게 앱 전역에서 특정한 목적 혹은 기능을 위해 사용될 수 있는 객체는 `Model`보다는 `Service`라는 대분류로 가져가는것이 좋아보입니다! `Service` - `Network` 이 하위에 이와 관련된 파일을 위치시키는게 맞지 않을까 싶어요. 

모델 `HTTPMethod` 같은 것들도 "네트워크 서비스"에서 사용되기에 이 하위로 가져가는게 좋아보입니다!

July911 committed:
refactor: 피드백 반영 #4

- JSONParsable 프로토콜 삭제
- 파일 분리
- Static 프로퍼티를 instance 프로퍼티로 변경
- httpMethod 열거형의 이름을 HTTPMethod로 변경

youngwoo-jang committed:
refactor: STEP1 피드백 반영 2차 #4

- Requester 객체가 decoding type을 스스로 정의하도록 JSONResponseDecodable 프로토콜 구현
- Requestable 프로토콜이 URLRequest가 아닌 HTTP메서드, URL등의 raw 정보를 요구하도록 변경
